src/routing/lazy_messaging.rs

```diff
@@ -6,121 +6,137 @@
     if section.prefix().matches(&src_name) {
         // This message is from our section. We update our members via the `Sync` message which is
         // done elsewhere.
-        return Ok(actions);
+        return Ok((actions, true));
     }
 
-    if let Ok(src_key) = msg.proof_chain_last_key() {
-        if !network.has_key(src_key) && !section.chain().has_key(src_key) {
-            // We don't know the src key. Send them a `OtherSection` message whose `dst_key` is set
-            // to the latest key we know and when they receive it they would send us back a
-            // `OtherSection` with their latest info, including their latest key.
-            //
-            // NOTE: the reason why we update the key only after receiving their `OtherSection` and
-            // not right here is to keep their key and their section_authority_provider in sync.
-            trace!("lazy messaging: src key unknown");
-            send_other_section = true;
+    let dst = msg.src().src_location().to_dst();
+
+    if let Ok(src_chain) = msg.proof_chain() {
+        if let Some(key) = network.key_by_name(&src_name) {
+            match src_chain.cmp_by_position(src_chain.last_key(), key) {
+                Ordering::Greater => {
+                    trace!("Anti-Entropy: We do not know source's key, need to update ourselves");
+                    let msg = create_other_section_message(
+                        node,
+                        section,
+                        network,
+                        Variant::SectionKnowledgeQuery {
+                            last_known_key: *key,
+                            msg: Box::new(msg.clone()),
+                        },
+                        dst,
+                    )?;
+                    actions.send.push(msg);
+                }
+                Ordering::Less => {
+                    info!("Anti-Entropy: Src is lagging and needs to update himself. Do nothing as he will update himself");
+                }
+                Ordering::Equal => {}
+            }
         }
     }
 
-    if let Some(new) = msg.dst_key() {
-        let src_prefix = network
-            .section_by_name(&src_name)
-            .1
-            .map(|info| &info.prefix);
-        if let Some(src_prefix) = src_prefix {
-            let old = network
-                .knowledge_by_name(&src_name)
-                .unwrap_or_else(|| section.chain().root_key());
-
-            if section.chain().cmp_by_position(new, old) == Ordering::Greater {
-                // Their knowledge of our section is newer than what we have stored - update it.
-                trace!("lazy messaging: dst key updated");
-                actions.propose = Some(Proposal::TheirKnowledge {
-                    prefix: *src_prefix,
-                    key: *new,
-                });
+    match section
+        .chain()
+        .cmp_by_position(&dest_info.dest_section_pk, section.chain().last_key())
+    {
+        Ordering::Greater => {
```

Comment:
this will never happen as `section.chain().last_key()` always has the highest index among its own chain